pulp_container/tests/functional/api/test_push_signatures.py

```diff
@@ -0,0 +1,151 @@
+"""Tests that verify that an image signature can be pushed to Pulp."""
+import gnupg
+import base64
+import json
+import requests
+import subprocess
+import pytest
+import tempfile
+
+from pulp_container.tests.functional.constants import REGISTRY_V2_REPO_PULP
+from pulp_container.constants import SIGNATURE_TYPE
+
+KEY = "https://raw.githubusercontent.com/pulp/pulp-fixtures/master/common/GPG-PRIVATE-KEY-pulp-qe"
+
+
+@pytest.fixture(scope="session")
+def tempdir():
+    """A temporary directory."""
+    with tempfile.TemporaryDirectory() as directory:
+        yield directory
+
+
+@pytest.fixture(scope="session")
+def gpg_metadata(tempdir):
+    """A fixture for a GPG instance and related metadata."""
+    private_key = requests.get(KEY).text
+
+    gpg = gnupg.GPG(gnupghome=tempdir)
+
+    gpg.import_keys(private_key)
+
+    fingerprint = gpg.list_keys()[0]["fingerprint"]
+    keyid = gpg.list_keys()[0]["keyid"]
+
+    gpg.trust_keys(fingerprint, "TRUST_ULTIMATE")
+
+    return gpg, fingerprint, keyid
+
+
+@pytest.fixture
+def distribution(
+    bindings_cfg,
+    registry_client,
+    local_registry,
+    container_distribution_api,
+    gpg_metadata,
+    add_to_cleanup,
+):
+    """Return a distribution created after pushing a signed content to the Pulp Registry."""
+    registry_client.pull(f"{REGISTRY_V2_REPO_PULP}:manifest_a")
+
+    image_path = f"{REGISTRY_V2_REPO_PULP}:manifest_a"
+    local_image_path = "/".join([local_registry.name, "test-1:manifest_a"])
+    gpg, fingerprint, keyid = gpg_metadata
+
+    registry_client.tag(image_path, local_image_path)
+    registry_client.login(
+        "-u", bindings_cfg.username, "-p", bindings_cfg.password, local_registry.name
+    )
+    subprocess.check_output(
+        [
+            "sudo",
+            "-E",
+            f"GNUPGHOME={gpg.gnupghome}",
+            "podman",
+            "push",
+            local_image_path,
+            "--sign-by",
+            keyid,
+        ]
+    )
+
+    # push another signature with a different timestamp
+    registry_client.tag(image_path, local_image_path)
+    registry_client.login(
+        "-u", bindings_cfg.username, "-p", bindings_cfg.password, local_registry.name
+    )
+    subprocess.check_output(
+        [
+            "sudo",
```

Comment:
Have you tried not to sudo at all?

Reply:
Yes. But, every other podman call is invoked with "sudo". So, I decided to stick to the standard. Besides that, I could easily set the environment variable with a GnuPG home directory.

Reply:
Oh, you mean using something like `monkeypatch.setenv(name, value, prepend=False)` from https://docs.pytest.org/en/6.2.x/monkeypatch.html to set the environment variable? :open_mouth: That looks interesting.

Reply:
This does not seem to be working:
```
    monkeypatch.setenv("GNUPGHOME", gpg.gnupghome)
    local_registry.push(image_path, "test-1:manifest_a", "--sign-by", keyid)
```
I can see the environment variable through `os.getenv("GNUPGHOME")`. No idea why podman cannot see the variable...

```
(pulp) [vagrant@pulp3-source-fedora35 ~]$ gpg --home /tmp/tmpbvcflazb --list-keys --keyid-format LONG
/tmp/tmpbvcflazb/pubring.kbx
----------------------------
pub   rsa2048/**05A5E6DA269D9D98** 2016-08-05 [SC]
      6EDF301256480B9B801EBA3D05A5E6DA269D9D98
uid                 [ultimate] Pulp QE
sub   rsa2048/E8DC6B59B70CEECB 2016-08-05 [E]
```
```
os.getenv("GNUPGHOME")
Out[11]: '/tmp/tmpbvcflazb'
local_registry.push(image_path, "test-1:manifest_a", "--sign-by", keyid)
Traceback (most recent call last):
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/IPython/core/interactiveshell.py", line 3369, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-12-c8e783671e35>", line 1, in <cell line: 1>
    local_registry.push(image_path, "test-1:manifest_a", "--sign-by", keyid)
  File "/home/vagrant/devel/pulp_container/pulp_container/tests/conftest.py", line 128, in push
    registry_client.push(local_image_path, *args)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pulp_smash/cli.py", line 1028, in _dispatch_command
    result = self._client.run(cmd)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pulp_smash/cli.py", line 320, in run
    return self.response_handler(completed_process)
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pulp_smash/cli.py", line 54, in code_handler
    completed_proc.check_returncode()
  File "/usr/local/lib/pulp/lib64/python3.10/site-packages/pulp_smash/cli.py", line 128, in check_returncode
    raise exceptions.CalledProcessError(
pulp_smash.exceptions.CalledProcessError: Command ('podman', 'push', 'pulp3-source-fedora35.localhost.example.com:443/test-1:manifest_a', '--sign-by', '05A5E6DA269D9D98') returned non-zero exit status 125.
```

Reply:
Okay, I need to do the following in order to set the environment variable:
```
registry_client._client.machine.env["GNUPGHOME"] = gpg.gnupghome
```
I am pushing the changes and merging the PR since it reached 45 comments.